File: writer.py
#coding: utf-8

import os
import web
import markdown
import hashlib

from chimchar import post, site_setting
from chimchar.db import reader
from chimchar.db import writer
from chimchar.utils import do_auth
from chimchar.utils import template_render

from config import SECRET

class auth_handler:
    def GET(self):
        site_info = reader.get_site_setting()
        if site_info == None:
            site_info = site_setting()
            site_info.site_name = 'demo'
            site_info.site_slogan = ''
            site_info.author = 'chimchar'
            site_info.google_analytic_id = ''
            writer.set_site_setting(site_info)
        site_info.site_domain = web.ctx.host
        template_values = {
                'site_info': site_info
                }
        path = os.path.join(os.path.dirname(__file__), 'tpl', 'writer')
        return template_render('auth.html', path, template_values)

    def POST(self):
        site_info = reader.get_site_setting()
        if site_info == None:
            site_info = site_setting()
            site_info.site_name = 'demo'
            site_info.site_slogan = ''
            site_info.author = 'chimchar'
            site_info.google_analytic_id = ''
            writer.set_site_setting(site_info)
        site_info.site_domain = web.ctx.host
        template_values = {
                'site_info': site_info
                }
        path = os.path.join(os.path.dirname(__file__), 'tpl', 'writer')

        pwd = web.input().passwd
        sha1 = hashlib.sha1(pwd).hexdigest()
        if (sha1 == SECRET):
            web.setcookie('is_pass', sha1, 3600)
            web.redirect('/overview')
        else:
            template_values['message'] = 'Secret pharse WRONG!'
            return template_render('auth.html', path, template_values)

class remove_handler(do_auth):
    def GET(self, id):
        post = reader.get_post_by_id(id)
        if post:
            writer.remove_post(post)
        web.redirect('/overview')
    def POST(self, id):
        post = reader.get_post_by_id(id)
        if post:
            writer.remove_post(post)
        web.redirect('/overview')

class overview_handler(do_auth):
    def GET(self):
        template_values = {
                'site_info': self.site_info,
                'posts': reader.get_all()
                }
        path = os.path.join(os.path.dirname(__file__), 'tpl', 'writer')
        return template_render('overview.html', path, template_values)

class write_handler(do_auth):
    def GET(self, id=''):
        template_values = {
                'site_info': self.site_info
                }
        path = os.path.join(os.path.dirname(__file__), 'tpl', 'writer')

        if (id):
            template_values['mode'] = 'edit'
            template_values['post'] = reader.get_post_by_id(id)
        else:
            template_values['mode'] = 'new'
            template_values['post'] = post()

        return template_render('write.html', path, template_values)

class index_write_handler(do_auth):
    def GET(self):
        template_values = {
                'site_info': self.site_info,
                'mode': 'edit',
                'post': reader.get_index()
                }
        path = os.path.join(os.path.dirname(__file__), 'tpl', 'writer')
        return template_render('index_write.html', path, template_values)

class index_update_handler(do_auth):
    def GET(self):
        web.redirect('/overview')

    def POST(self):
        p = web.input()
        writer.set_index(p)
        web.redirect('/overview')

class post_handler(do_auth):
    def GET(self, id=''):
        web.redirect('/overview')

    def POST(self, id=''):
        p = web.input()
        p.id = id
        writer.set_post_by_id(p)
        web.redirect('/overview')

class settings_handler(do_auth):
    def GET(self):
        template_values = {
                'site_info': self.site_info
                }
        path = os.path.join(os.path.dirname(__file__), 'tpl', 'writer')
        return template_render('settings.html', path, template_values)

    def POST(self):
        s = web.input()
        writer.set_site_setting(s)
        web.redirect('/overview')

urls = (
        '/auth', 'auth_handler',
        '/overview', 'overview_handler',
        '/settings', 'settings_handler',
        '/new', 'write_handler',
        '/save', 'post_handler',
        '/edit/index', 'index_write_handler',
        '/update/index', 'index_update_handler',
        '/edit/([0-9a-zA-Z\-\_]+)', 'write_handler',
        '/update/([0-9a-zA-Z\-\_]+)', 'post_handler',
        '/remove/([0-9a-zA-Z\-\_]+)', 'remove_handler',
        '', 'overview_handler',
        '/', 'overview_handler',
        )

web.config.debug = False
app = web.application(urls, locals())

File: writer.php
<?php if (!defined('BASEPATH')) exit('No direct sciprt access!');

class auth {
    function __construct() {
        include('secret'.EXT);
        $this->secret = $secret;
        
        $param = array(
            'key' => 'chimchar',
            'sess_name' => 'chimchar',
            'lifetime' => '3600',
        );
        $this->session = Charizard::load('sessioner', $param);
        $this->status = Charizard::load('status_coder');
    }

    function check($key) {
        if (sha1($key) === $this->secret) {
            return true;
        } else {
            return false;
        }
    }

    function set() {
        $this->session->userdata['is_pass'] = true;
        $this->session->sess_write();
    }

    function kill() {
        $this->session->sess_destory();
        $this->session = array();
    }

    function do_auth($key) {
        if ($this->is_pass()) {
            return true;
        }

        if ($this->check($key)) {
            $this->set();
            return true;
        }

        $this->kill();
        return false;
    }

    function is_pass() {
        return $this->session->userdata['is_pass'];
    }

    function back_auth($message = false) {
        $u = Charizard::load('url_helper');
        $this->status->redirect($u->build('/writer/login', $message));
    }
}

class login extends auth {
    function GET() {
        if ($this->is_pass()) {
            $u = Charizard::load('url_helper');
            $this->status->redirect($u->build('/writer/overview'));
        }

        $render = Charizard::load('render');
        $template_values = array(
            'title' => 'login',
            'u' => Charizard::load('url_helper')
        );

        $path = 'writer';
        echo $render->rende('auth.html', $path, $template_values);
    }

    function POST() {
        if ($this->do_auth($_POST['passwd'])) {
            $u = Charizard::load('url_helper');
            $this->status->seeother($u->build('/writer/overview'));
        } else {
            $message = array(
                'message' => 'Secret parse wrong!',
                'message' => false,
                'message' => false
            );
            $this->back_auth($message);
        }
    }
}

class logout extends auth {
    public $message = false;

    function GET() {
        if (!$this->is_pass()) {
            $message = array(
                'message' => 'login first',
            );
            $this->back_auth($message);
        } else {
            $this->kill();
            $message = array(
                'message' => 'Bye my friend =)'
            );
            $this->back_auth($message);
        }
    }
}

class remove_post_handler extends auth {
    function GET($id) {
        if (!$this->is_pass()) {
            $message = array(
                'message' => 'login first',
            );
            $this->back_auth($message);
            return;
        }

        $db = Charizard::load('db');
        $post = $db->get_post_by_id($id);
        if ($post) {
            $db->remove_post($id);
            $message = array(
                'message' => "post $id remove.",
            );
        } else {
            $message = array(
                'message' => "post $id not found!",
            );
        }
        $u = Charizard::load('url_helper');
        $this->status->redirect($u->build('/writer/overview', $message));
    }

    function POST($id) {
        $this->GET($id);
    }
}

class remove_minisite_handler extends auth {
    function GET($id) {
        if (!$this->is_pass()) {
            $message = array(
                'message' => 'login first',
            );
            $this->back_auth($message);
            return;
        }

        $db = Charizard::load('db');
        $post = $db->get_minisite_by_id($id);
        if ($post) {
            $db->remove_minisite($id);
            $message = array(
                'message' => "minisite $id remove.",
            );
        } else {
            $message = array(
                'message' => "minisite $id not found!",
            );
        }
        $u = Charizard::load('url_helper');
        $this->status->redirect($u->build('/writer/overview', $message));
    }

    function POST($id) {
        $this->GET($id);
    }
}

class overview_handler extends auth {
    function GET() {
        if (!$this->is_pass()) {
            $message = array(
                'message' => 'login first',
            );
            $this->back_auth($message);
            return;
        }

        $db = Charizard::load('db');
        $render = Charizard::load('render');

        $posts = $db->get_posts();
        $minisites = $db->get_minisites();
        rsort($posts);
        rsort($minisites);
        $template_values = array(
            'site_info' => $db->get_site(),
            'u' => Charizard::load('url_helper'),
            'posts' => $posts,
            'minisites' => $minisites
        );

        $path = 'writer';
        echo $render->rende('overview.html', $path, $template_values);
    }
}

class write_post_handler extends auth {
    function GET($id = false) {
        if (!$this->is_pass()) {
            $message = array(
                'message' => 'login first',
            );
            $this->back_auth($message);
            return;
        }

        $db = Charizard::load('db');
        $render = Charizard::load('render');

        $template_values = array(
            'site_info' => $db->get_site(),
            'u' => Charizard::load('url_helper'),
        );

        if ($id) {
            $template_values['mode'] = 'edit';
            $template_values['post'] = $db->get_post_by_id($id);
        } else {
            $template_values['mode'] = 'new';
            $template_values['post'] = $db->get_new('posts');
        }

        $path = 'writer';
        echo $render->rende('write_post.html', $path, $template_values);
    }
}

class write_minisite_handler extends auth {
    function GET($id = false) {
        if (!$this->is_pass()) {
            $message = array(
                'message' => 'login first',
            );
            $this->back_auth($message);
            return;
        }

        $db = Charizard::load('db');
        $render = Charizard::load('render');

        $template_values = array(
            'site_info' => $db->get_site(),
            'u' => Charizard::load('url_helper'),
        );

        if ($id) {
            $template_values['mode'] = 'edit';
            $template_values['post'] = $db->get_minisite_by_id($id);
        } else {
            $template_values['mode'] = 'new';
            $template_values['post'] = $db->get_new('minisite');
        }

        $path = 'writer';
        echo $render->rende('write_minisite.html', $path, $template_values);
    }
}
class write_index_handler extends auth {
    function GET() {
        if (!$this->is_pass()) {
            $message = array(
                'message' => 'login first',
            );
            $this->back_auth($message);
            return;
        }

        $db = Charizard::load('db');
        $render = Charizard::load('render');

        $template_values = array(
            'site_info' => $db->get_site(),
            'mode' => 'edit',
            'post' => $db->get_index(),
            'u' => Charizard::load('url_helper'),
        );

        $path = 'writer';
        echo $render->rende('write_index.html', $path, $template_values);
    }
}

class update_index_handler extends auth {
    function POST() {
        $db = Charizard::load('db');
        $status_coder = Charizard::load('status_coder');
        $u = Charizard::load('url_helper');
        $mkd = Charizard::load('mkd_render');

        $p = $_POST;
        $p['formatted_content'] = $mkd->rende($p['content']);
        $db->set_index($p);

        $message = array(
            'message' => 'Your post saved',
        );

        $status_coder->_301($u->build('/writer/overview'));
    }
}

class update_post_handler extends auth {
    function POST($id = false) {
        $db = Charizard::load('db');
        $status_coder = Charizard::load('status_coder');
        $u = Charizard::load('url_helper');
        $mkd = Charizard::load('mkd_render');

        $p = $_POST;
        $p['id'] = $id;
        $p['formatted_content'] = $mkd->rende($p['content']);
        $p['modified_time'] = date(DATE_FORMAT);
        if (!$id) {
            $p['created_time'] = date(DATE_FORMAT);
        }
        $db->set_post($p);

        $message = array(
            'message' => 'Your post saved',
        );

        $status_coder->_301($u->build('/writer/overview', $message));
    }
}

class update_minisite_handler extends auth {
    function POST($id = false) {
        $db = Charizard::load('db');
        $status_coder = Charizard::load('status_coder');
        $u = Charizard::load('url_helper');
        $mkd = Charizard::load('mkd_render');

        $p = $_POST;
        $p['id'] = $id;
        $p['formatted_content'] = $mkd->rende($p['content']);
        $p['modified_time'] = date(DATE_FORMAT);
        if (!$id) {
            $p['created_time'] = date(DATE_FORMAT);
        }
        $db->set_minisite($p);

        $message = array(
            'message' => 'Your minisite saved',
        );

        $status_coder->_301($u->build('/writer/overview', $message));
    }
}

class settings_handler extends auth {
    function GET() {
        if (!$this->is_pass()) {
            $message = array(
                'message' => 'login first',
            );
            $this->back_auth($message);
            return;
        }

        $db = Charizard::load('db');
        $render = Charizard::load('render');
        
        $template_values = array(
            'site_info' => $db->get_site(),
            'u' => Charizard::load('url_helper'),
        );

        $path = 'writer';
        echo $render->rende('settings.html', $path, $template_values);
    }

    function POST() {
        $db = Charizard::load('db');
        $status_coder = Charizard::load('status_coder');
        $u = Charizard::load('url_helper');
    
        $s = $_POST;
        $db->set_site($s);

        $message = array(
            'message' => 'Your settings saved',
        );

        $status_coder->_301($u->build('/writer/overview', $message));
    }
}
?>
